docs/middleware.rst

```diff
@@ -40,18 +40,78 @@ Pythonic
     where appropriate. For example, it converts the raw hex string returned by the RPC call
     ``eth_blockNumber`` into an ``int``.
 
-Built-in Middleware
+.. _Modifying_Middleware:
+
+Modifying Middleware
 ------------------
 
-Web3 ships with middleware for custom use, as desired. Middleware can be added after creating
-your Web3 object, like:
+Middleware can be added, removed, replaced, and cleared at runtime. To make that easier, you
+can name the middleware for later reference. Alternatively, you can use a reference to the
+middleware itself.
 
-.. code-block:: python
+.. py:method:: Web3.middleware_stack.add(middleware, name=None)
+
+    Middleware will be added to the top of the stack. That means the new middleware will modify the
+    request first, and the response last. You can optionally name it with any hashable object,
+    typically a string.
+
+    .. code-block:: python
+
+        >>> w3 = Web3(...)
```

Comment:
I don't love `web3` being the instance name actually, since the package is also called web3. But consistency is more important here, so I agree in the context of this PR and will update. Later, I'll see if I can convince myself (and then you) that we should change the standard practice everywhere in the docs and tests. It's a big change now, but will only get more massive over time.

For now, I'll go with something like:

```
from web3.middleware import pythonic_middleware
web3 = Web3(...)
web3.middleware_stack.add(web3.middleware.pythonic_middleware)
```

(A little preview of my raw thoughts: the current standard means you can never use `import web3`, which is a valid style choice in some repositories. You end up losing access to the module as soon as you initialize `web3 = Web3()`. I was using the module that way in the little snippets, with `web3.middleware.pythonic_middleware`, for example. The standard also probably leads to confusion for beginners who might expect to `import web3` and use it statically without initializing.)

Reply:
I'm convinced.  `w3` works for me.